File: functions/reminders-service.js
/**
 * Reminders service - Business logic for querying and filtering reminders
 */

import { getFirestore } from './firebase-config.js';
import {
  sortByDueDate,
  sortByPriority,
  filterByStatus,
  searchReminders,
  formatReminderForLLM,
  daysUntil,
  isToday,
  isPast,
  isOverdue,
  getStatus,
} from './utils.js';

/**
 * Get all reminders for a user
 * @param {string} userId - User email or ID
 * @param {Object} options - Query options
 * @returns {Promise<Array>}
 */
export async function getAllReminders(userId, options = {}) {
  const {
    status = 'all',
    limit = null,
    formatForLLM = true,
  } = options;

  try {
    const db = getFirestore();
    let query = db.collection('reminders').where('ownerId', '==', userId);

    // Filter by completion status
    if (status === 'completed') {
      query = query.where('isCompleted', '==', true);
    } else if (status === 'pending') {
      query = query.where('isCompleted', '==', false);
    }

    const snapshot = await query.get();
    let reminders = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data(),
    }));

    // Apply limit if specified
    if (limit) {
      reminders = reminders.slice(0, limit);
    }

    // Format for LLM consumption if requested
    if (formatForLLM) {
      reminders = reminders.map(formatReminderForLLM);
    }

    return reminders;
  } catch (error) {
    console.error('Error fetching reminders:', error);
    throw error;
  }
}

/**
 * Get upcoming reminders for the next N days
 * @param {string} userId - User email or ID
 * @param {number} days - Number of days to look ahead (default: 7)
 * @param {Object} options - Query options
 * @returns {Promise<Array>}
 */
export async function getUpcomingReminders(userId, days = 7, options = {}) {
  const {
    sortBy = 'dueDate',
    includeCompleted = false,
    formatForLLM = true,
  } = options;

  try {
    const db = getFirestore();
    const now = new Date();
    const futureDate = new Date();
    futureDate.setDate(futureDate.getDate() + days);

    let query = db.collection('reminders')
      .where('ownerId', '==', userId)
      .where('dueAt', '>=', now)
      .where('dueAt', '<=', futureDate);

    if (!includeCompleted) {
      query = query.where('isCompleted', '==', false);
    }

    const snapshot = await query.get();
    let reminders = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data(),
    }));

    // Sort results
    if (sortBy === 'dueDate') {
      reminders = sortByDueDate(reminders, 'asc');
    } else if (sortBy === 'priority') {
      reminders = sortByPriority(reminders);
    }

    // Format for LLM consumption if requested
    if (formatForLLM) {
      reminders = reminders.map(formatReminderForLLM);
    }

    return reminders;
  } catch (error) {
    console.error('Error fetching upcoming reminders:', error);
    throw error;
  }
}

/**
 * Get reminders due today
 * @param {string} userId - User email or ID
 * @param {Object} options - Query options
 * @returns {Promise<Array>}
 */
export async function getTodayReminders(userId, options = {}) {
  const {
    includeCompleted = false,
    formatForLLM = true,
  } = options;

  try {
    const db = getFirestore();
    const reminders = await getAllReminders(userId, {
      status: includeCompleted ? 'all' : 'pending',
      formatForLLM: false,
    });

    const todayReminders = reminders.filter(r =>
      r.dueAt && isToday(new Date(r.dueAt))
    );

    // Format for LLM consumption if requested
    if (formatForLLM) {
      return todayReminders.map(formatReminderForLLM);
    }

    return todayReminders;
  } catch (error) {
    console.error('Error fetching today reminders:', error);
    throw error;
  }
}

/**
 * Get overdue reminders
 * @param {string} userId - User email or ID
 * @param {Object} options - Query options
 * @returns {Promise<Array>}
 */
export async function getOverdueReminders(userId, options = {}) {
  const {
    formatForLLM = true,
  } = options;

  try {
    const reminders = await getAllReminders(userId, {
      status: 'pending',
      formatForLLM: false,
    });

    const overdueReminders = reminders.filter(r =>
      r.dueAt && isPast(new Date(r.dueAt))
    );

    // Format for LLM consumption if requested
    if (formatForLLM) {
      return overdueReminders.map(formatReminderForLLM);
    }

    return overdueReminders;
  } catch (error) {
    console.error('Error fetching overdue reminders:', error);
    throw error;
  }
}

/**
 * Get a specific reminder by ID
 * @param {string} reminderId - Reminder ID
 * @param {string} userId - User email or ID (for validation)
 * @param {Object} options - Query options
 * @returns {Promise<Object>}
 */
export async function getReminderById(reminderId, userId, options = {}) {
  const { formatForLLM = true } = options;

  try {
    const db = getFirestore();
    const doc = await db.collection('reminders').doc(reminderId).get();

    if (!doc.exists) {
      throw new Error('Reminder not found');
    }

    const reminder = {
      id: doc.id,
      ...doc.data(),
    };

    // Verify ownership
    if (reminder.ownerId !== userId && 
        (!reminder.sharedWith || !reminder.sharedWith.includes(userId))) {
      throw new Error('Access denied');
    }

    if (formatForLLM) {
      return formatReminderForLLM(reminder);
    }

    return reminder;
  } catch (error) {
    console.error('Error fetching reminder:', error);
    throw error;
  }
}

/**
 * Search reminders
 * @param {string} userId - User email or ID
 * @param {string} query - Search query
 * @param {Object} options - Query options
 * @returns {Promise<Array>}
 */
export async function searchUserReminders(userId, query, options = {}) {
  const {
    status = 'all',
    limit = null,
    formatForLLM = true,
  } = options;

  try {
    let reminders = await getAllReminders(userId, {
      status,
      formatForLLM: false,
    });

    reminders = searchReminders(reminders, query);

    if (limit) {
      reminders = reminders.slice(0, limit);
    }

    if (formatForLLM) {
      return reminders.map(formatReminderForLLM);
    }

    return reminders;
  } catch (error) {
    console.error('Error searching reminders:', error);
    throw error;
  }
}

/**
 * Get reminders summary for a user
 * @param {string} userId - User email or ID
 * @returns {Promise<Object>}
 */
export async function getRemindersSummary(userId) {
  try {
    const reminders = await getAllReminders(userId, {
      status: 'all',
      formatForLLM: false,
    });

    const completed = reminders.filter(r => r.isCompleted);
    const pending = reminders.filter(r => !r.isCompleted);
    const overdue = pending.filter(r => r.dueAt && isPast(new Date(r.dueAt)));
    const dueToday = reminders.filter(r =>
      r.dueAt && isToday(new Date(r.dueAt)) && !r.isCompleted
    );
    const upcoming = pending.filter(r =>
      r.dueAt && !isPast(new Date(r.dueAt))
    );

    return {
      total: reminders.length,
      completed: completed.length,
      pending: pending.length,
      overdue: overdue.length,
      dueToday: dueToday.length,
      upcoming: upcoming.length,
      completionRate: reminders.length > 0 
        ? ((completed.length / reminders.length) * 100).toFixed(1)
        : 0,
    };
  } catch (error) {
    console.error('Error generating summary:', error);
    throw error;
  }
}

/**
 * Get shared reminders for a user
 * @param {string} userId - User email or ID
 * @param {Object} options - Query options
 * @returns {Promise<Array>}
 */
export async function getSharedReminders(userId, options = {}) {
  const {
    formatForLLM = true,
  } = options;

  try {
    const db = getFirestore();
    const snapshot = await db.collection('reminders')
      .where('sharedWith', 'array-contains', userId)
      .get();

    let reminders = snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data(),
    }));

    if (formatForLLM) {
      return reminders.map(formatReminderForLLM);
    }

    return reminders;
  } catch (error) {
    console.error('Error fetching shared reminders:', error);
    throw error;
  }
}

/**
 * Create a new reminder
 * @param {string} userId - User email or ID (owner)
 * @param {string} title - Reminder title
 * @param {Object} options - Reminder options
 * @returns {Promise<Object>}
 */
export async function createReminder(userId, title, options = {}) {
  const {
    notes = null,
    dueAt = null,
    recurrence = null,
    remindBeforeMinutes = 10,
    recurrenceEndDate = null,
    weeklyDays = null,
    sharedWith = null,
    formatForLLM = true,
  } = options;

  try {
    if (!title || title.trim().length === 0) {
      throw new Error('Title is required');
    }

    const db = getFirestore();
    const now = new Date();

    const reminderData = {
      title: title.trim(),
      notes: notes ? notes.trim() : null,
      createdAt: now,
      updatedAt: now,
      ownerId: userId,
      isCompleted: false,
      completedAt: null,
      isDisabled: false,
      disabledUntil: null,
      remindBeforeMinutes,
      dueAt: dueAt ? new Date(dueAt) : null,
      recurrence: recurrence || null,
      recurrenceEndDate: recurrenceEndDate ? new Date(recurrenceEndDate) : null,
      weeklyDays: weeklyDays || null,
      isShared: sharedWith && sharedWith.length > 0,
      sharedWith: sharedWith || [],
      lastModifiedBy: userId,
      deleted: false,
      version: 1,
    };

    const docRef = await db.collection('reminders').add(reminderData);

    const reminder = {
      id: docRef.id,
      ...reminderData,
    };

    if (formatForLLM) {
      return formatReminderForLLM(reminder);
    }

    return reminder;
  } catch (error) {
    console.error('Error creating reminder:', error);
    throw error;
  }
}

/**
 * Update an existing reminder
 * @param {string} reminderId - Reminder ID
 * @param {string} userId - User email or ID (must be owner)
 * @param {Object} updates - Fields to update
 * @returns {Promise<Object>}
 */
export async function updateReminder(reminderId, userId, updates = {}) {
  const { formatForLLM = true } = updates;
  const updateFields = { ...updates };
  delete updateFields.formatForLLM;

  try {
    const db = getFirestore();
    const reminderRef = db.collection('reminders').doc(reminderId);
    const doc = await reminderRef.get();

    if (!doc.exists) {
      throw new Error('Reminder not found');
    }

    const reminder = doc.data();

    // Verify ownership
    if (reminder.ownerId !== userId) {
      throw new Error('Access denied: You can only edit your own reminders');
    }

    // Validate title if being updated
    if (updateFields.title !== undefined && 
        (!updateFields.title || updateFields.title.trim().length === 0)) {
      throw new Error('Title cannot be empty');
    }

    // Prepare update data
    const dataToUpdate = {
      ...updateFields,
      updatedAt: new Date(),
      lastModifiedBy: userId,
      version: (reminder.version || 0) + 1,
    };

    // Convert dates if provided
    if (updateFields.dueAt !== undefined) {
      dataToUpdate.dueAt = updateFields.dueAt ? new Date(updateFields.dueAt) : null;
    }
    if (updateFields.recurrenceEndDate !== undefined) {
      dataToUpdate.recurrenceEndDate = updateFields.recurrenceEndDate 
        ? new Date(updateFields.recurrenceEndDate) 
        : null;
    }

    // Update shared status based on sharedWith array
    if (updateFields.sharedWith !== undefined) {
      dataToUpdate.isShared = updateFields.sharedWith && updateFields.sharedWith.length > 0;
    }

    await reminderRef.update(dataToUpdate);

    const updatedReminder = {
      id: reminderId,
      ...reminder,
      ...dataToUpdate,
    };

    if (formatForLLM) {
      return formatReminderForLLM(updatedReminder);
    }

    return updatedReminder;
  } catch (error) {
    console.error('Error updating reminder:', error);
    throw error;
  }
}

/**
 * Delete a reminder
 * @param {string} reminderId - Reminder ID
 * @param {string} userId - User email or ID (must be owner)
 * @returns {Promise<Object>}
 */
export async function deleteReminder(reminderId, userId) {
  try {
    const db = getFirestore();
    const reminderRef = db.collection('reminders').doc(reminderId);
    const doc = await reminderRef.get();

    if (!doc.exists) {
      throw new Error('Reminder not found');
    }

    const reminder = doc.data();

    // Verify ownership
    if (reminder.ownerId !== userId) {
      throw new Error('Access denied: You can only delete your own reminders');
    }

    // Soft delete - mark as deleted instead of removing
    await reminderRef.update({
      deleted: true,
      deletedAt: new Date(),
      lastModifiedBy: userId,
      version: (reminder.version || 0) + 1,
    });

    return {
      success: true,
      message: `Reminder "${reminder.title}" deleted successfully`,
      reminderId,
    };
  } catch (error) {
    console.error('Error deleting reminder:', error);
    throw error;
  }
}

/**
 * Mark a reminder as completed
 * @param {string} reminderId - Reminder ID
 * @param {string} userId - User email or ID (must be owner or has access)
 * @returns {Promise<Object>}
 */
export async function completeReminder(reminderId, userId) {
  try {
    const db = getFirestore();
    const reminderRef = db.collection('reminders').doc(reminderId);
    const doc = await reminderRef.get();

    if (!doc.exists) {
      throw new Error('Reminder not found');
    }

    const reminder = doc.data();

    // Verify access (owner or shared with user)
    if (reminder.ownerId !== userId && 
        (!reminder.sharedWith || !reminder.sharedWith.includes(userId))) {
      throw new Error('Access denied: You do not have permission to complete this reminder');
    }

    const now = new Date();

    await reminderRef.update({
      isCompleted: true,
      completedAt: now,
      lastModifiedBy: userId,
      version: (reminder.version || 0) + 1,
    });

    const completedReminder = {
      id: reminderId,
      ...reminder,
      isCompleted: true,
      completedAt: now,
      lastModifiedBy: userId,
      version: (reminder.version || 0) + 1,
    };

    return formatReminderForLLM(completedReminder);
  } catch (error) {
    console.error('Error completing reminder:', error);
    throw error;
  }
}

File: mcp/python/reminders_service.py
"""
Reminders service - Business logic for querying and filtering reminders
"""

from datetime import datetime
from typing import List, Dict, Any, Optional
from firebase_config import get_firestore
from utils import (
    sort_by_due_date,
    sort_by_priority,
    filter_by_status,
    search_reminders,
    format_reminder_for_llm,
    days_until,
    is_today,
    is_past,
    is_overdue,
    get_status,
)


async def get_all_reminders(
    user_id: str,
    status: str = "all",
    limit: Optional[int] = None,
    format_for_llm: bool = True,
) -> List[Dict[str, Any]]:
    """
    Get all reminders for a user
    
    Args:
        user_id: User email or ID
        status: 'pending', 'completed', or 'all'
        limit: Maximum number of reminders
        format_for_llm: Format for LLM consumption
        
    Returns:
        List of reminders
    """
    try:
        db = get_firestore()
        query = db.collection("shared_tasks").where("ownerId", "==", user_id)
        
        # Filter by completion status
        if status == "completed":
            query = query.where("isCompleted", "==", True)
        elif status == "pending":
            query = query.where("isCompleted", "==", False)
        
        docs = query.stream()
        reminders = [
            {
                "id": doc.id,
                **doc.to_dict(),
            }
            for doc in docs
        ]
        
        # Apply limit if specified
        if limit:
            reminders = reminders[:limit]
        
        # Format for LLM consumption if requested
        if format_for_llm:
            reminders = [format_reminder_for_llm(r) for r in reminders]
        
        return reminders
    except Exception as error:
        raise Exception(f"Error fetching reminders: {str(error)}")


async def get_upcoming_reminders(
    user_id: str,
    days: int = 7,
    sort_by: str = "priority",
    include_completed: bool = False,
    format_for_llm: bool = True,
) -> List[Dict[str, Any]]:
    """
    Get upcoming reminders for the next N days
    
    Args:
        user_id: User email or ID
        days: Number of days to look ahead
        sort_by: 'dueDate' or 'priority'
        include_completed: Include completed reminders
        format_for_llm: Format for LLM consumption
        
    Returns:
        List of upcoming reminders
    """
    try:
        db = get_firestore()
        now = datetime.now()
        future_date = datetime.fromtimestamp(
            now.timestamp() + (days * 24 * 3600)
        )
        
        query = (
            db.collection("shared_tasks")
            .where("ownerId", "==", user_id)
            .where("dueAt", ">=", now)
            .where("dueAt", "<=", future_date)
        )
        
        if not include_completed:
            query = query.where("isCompleted", "==", False)
        
        docs = query.stream()
        reminders = [
            {
                "id": doc.id,
                **doc.to_dict(),
            }
            for doc in docs
        ]
        
        # Sort results
        if sort_by == "dueDate":
            reminders = sort_by_due_date(reminders, "asc")
        elif sort_by == "priority":
            reminders = sort_by_priority(reminders)
        
        # Format for LLM consumption if requested
        if format_for_llm:
            reminders = [format_reminder_for_llm(r) for r in reminders]
        
        return reminders
    except Exception as error:
        raise Exception(f"Error fetching upcoming reminders: {str(error)}")


async def get_today_reminders(
    user_id: str,
    include_completed: bool = False,
    format_for_llm: bool = True,
) -> List[Dict[str, Any]]:
    """
    Get reminders due today
    
    Args:
        user_id: User email or ID
        include_completed: Include completed reminders
        format_for_llm: Format for LLM consumption
        
    Returns:
        List of today's reminders
    """
    try:
        reminders = await get_all_reminders(
            user_id,
            status="all" if include_completed else "pending",
            format_for_llm=False,
        )
        
        today_reminders = [r for r in reminders if is_today(r.get("dueAt"))]
        
        # Format for LLM consumption if requested
        if format_for_llm:
            today_reminders = [format_reminder_for_llm(r) for r in today_reminders]
        
        return today_reminders
    except Exception as error:
        raise Exception(f"Error fetching today reminders: {str(error)}")


async def get_overdue_reminders(
    user_id: str,
    format_for_llm: bool = True,
) -> List[Dict[str, Any]]:
    """
    Get overdue reminders
    
    Args:
        user_id: User email or ID
        format_for_llm: Format for LLM consumption
        
    Returns:
        List of overdue reminders
    """
    try:
        reminders = await get_all_reminders(
            user_id,
            status="pending",
            format_for_llm=False,
        )
        
        overdue_reminders = [r for r in reminders if is_overdue(r)]
        
        # Format for LLM consumption if requested
        if format_for_llm:
            overdue_reminders = [format_reminder_for_llm(r) for r in overdue_reminders]
        
        return overdue_reminders
    except Exception as error:
        raise Exception(f"Error fetching overdue reminders: {str(error)}")


async def get_reminder_by_id(
    reminder_id: str,
    user_id: str,
    format_for_llm: bool = True,
) -> Dict[str, Any]:
    """
    Get a specific reminder by ID
    
    Args:
        reminder_id: Reminder ID
        user_id: User email or ID (for validation)
        format_for_llm: Format for LLM consumption
        
    Returns:
        Reminder object
    """
    try:
        db = get_firestore()
        doc = db.collection("shared_tasks").document(reminder_id).get()
        
        if not doc.exists:
            raise Exception("Reminder not found")
        
        reminder = {
            "id": doc.id,
            **doc.to_dict(),
        }
        
        # Verify ownership
        if (
            reminder.get("ownerId") != user_id
            and user_id not in (reminder.get("sharedWith") or [])
        ):
            raise Exception("Access denied")
        
        if format_for_llm:
            return format_reminder_for_llm(reminder)
        
        return reminder
    except Exception as error:
        raise Exception(f"Error fetching reminder: {str(error)}")


async def search_user_reminders(
    user_id: str,
    query: str,
    status: str = "all",
    limit: Optional[int] = None,
    format_for_llm: bool = True,
) -> List[Dict[str, Any]]:
    """
    Search reminders
    
    Args:
        user_id: User email or ID
        query: Search query
        status: 'pending', 'completed', or 'all'
        limit: Maximum results
        format_for_llm: Format for LLM consumption
        
    Returns:
        List of matching reminders
    """
    try:
        reminders = await get_all_reminders(
            user_id,
            status=status,
            format_for_llm=False,
        )
        
        reminders = search_reminders(reminders, query)
        
        if limit:
            reminders = reminders[:limit]
        
        if format_for_llm:
            reminders = [format_reminder_for_llm(r) for r in reminders]
        
        return reminders
    except Exception as error:
        raise Exception(f"Error searching reminders: {str(error)}")


async def get_reminders_summary(user_id: str) -> Dict[str, Any]:
    """
    Get reminders summary for a user
    
    Args:
        user_id: User email or ID
        
    Returns:
        Summary statistics
    """
    try:
        reminders = await get_all_reminders(
            user_id,
            status="all",
            format_for_llm=False,
        )
        
        completed = [r for r in reminders if r.get("isCompleted")]
        pending = [r for r in reminders if not r.get("isCompleted")]
        overdue = [r for r in pending if r.get("dueAt") and is_past(r["dueAt"])]
        due_today = [
            r for r in reminders
            if r.get("dueAt") and is_today(r["dueAt"]) and not r.get("isCompleted")
        ]
        upcoming = [r for r in pending if r.get("dueAt") and not is_past(r["dueAt"])]
        
        return {
            "total": len(reminders),
            "completed": len(completed),
            "pending": len(pending),
            "overdue": len(overdue),
            "dueToday": len(due_today),
            "upcoming": len(upcoming),
            "completionRate": (
                round((len(completed) / len(reminders)) * 100, 1) if reminders else 0
            ),
        }
    except Exception as error:
        raise Exception(f"Error generating summary: {str(error)}")


async def get_shared_reminders(
    user_id: str,
    format_for_llm: bool = True,
) -> List[Dict[str, Any]]:
    """
    Get shared reminders for a user
    
    Args:
        user_id: User email or ID
        format_for_llm: Format for LLM consumption
        
    Returns:
        List of shared reminders
    """
    try:
        db = get_firestore()
        docs = db.collection("shared_tasks").where(
            "sharedWith", "array-contains", user_id
        ).stream()
        
        reminders = [
            {
                "id": doc.id,
                **doc.to_dict(),
            }
            for doc in docs
        ]
        
        if format_for_llm:
            reminders = [format_reminder_for_llm(r) for r in reminders]
        
        return reminders
    except Exception as error:
        raise Exception(f"Error fetching shared reminders: {str(error)}")


async def create_reminder(
    user_id: str,
    title: str,
    notes: Optional[str] = None,
    due_at: Optional[str] = None,
    recurrence: Optional[str] = None,
    remind_before_minutes: int = 10,
    recurrence_end_date: Optional[str] = None,
    weekly_days: Optional[List[int]] = None,
    shared_with: Optional[List[str]] = None,
    format_for_llm: bool = True,
) -> Dict[str, Any]:
    """
    Create a new reminder
    
    Args:
        user_id: User email or ID (owner)
        title: Reminder title
        notes: Optional notes
        due_at: Due date/time in ISO format
        recurrence: Recurrence pattern
        remind_before_minutes: Minutes before to remind
        recurrence_end_date: Recurrence end date
        weekly_days: Weekly recurrence days
        shared_with: List of emails to share with
        format_for_llm: Format for LLM consumption
        
    Returns:
        Created reminder object
    """
    try:
        if not title or not title.strip():
            raise ValueError("Title is required")
        
        db = get_firestore()
        now = datetime.now()
        
        reminder_data = {
            "title": title.strip(),
            "notes": notes.strip() if notes else None,
            "createdAt": now,
            "updatedAt": now,
            "ownerId": user_id,
            "isCompleted": False,
            "completedAt": None,
            "isDisabled": False,
            "disabledUntil": None,
            "remindBeforeMinutes": remind_before_minutes,
            "dueAt": datetime.fromisoformat(due_at.replace("Z", "+00:00")) if due_at else None,
            "recurrence": recurrence or None,
            "recurrenceEndDate": (
                datetime.fromisoformat(recurrence_end_date.replace("Z", "+00:00"))
                if recurrence_end_date
                else None
            ),
            "weeklyDays": weekly_days or None,
            "isShared": bool(shared_with and len(shared_with) > 0),
            "sharedWith": shared_with or [],
            "lastModifiedBy": user_id,
            "deleted": False,
            "version": 1,
        }
        
        doc_ref = db.collection("shared_tasks").add(reminder_data)
        
        reminder = {
            "id": doc_ref[1].id,
            **reminder_data,
        }
        
        if format_for_llm:
            return format_reminder_for_llm(reminder)
        
        return reminder
    except Exception as error:
        raise Exception(f"Error creating reminder: {str(error)}")


async def update_reminder(
    reminder_id: str,
    user_id: str,
    **updates,
) -> Dict[str, Any]:
    """
    Update an existing reminder
    
    Args:
        reminder_id: Reminder ID
        user_id: User email or ID (must be owner)
        **updates: Fields to update
        
    Returns:
        Updated reminder object
    """
    try:
        format_for_llm = updates.pop("format_for_llm", True)
        
        db = get_firestore()
        reminder_ref = db.collection("shared_tasks").document(reminder_id)
        doc = reminder_ref.get()
        
        if not doc.exists:
            raise Exception("Reminder not found")
        
        reminder = doc.to_dict()
        
        # Verify ownership
        if reminder.get("ownerId") != user_id:
            raise Exception("Access denied: You can only edit your own reminders")
        
        # Validate title if being updated
        if "title" in updates and (not updates["title"] or not updates["title"].strip()):
            raise Exception("Title cannot be empty")
        
        # Prepare update data
        update_data = {
            **updates,
            "updatedAt": datetime.now(),
            "lastModifiedBy": user_id,
            "version": (reminder.get("version") or 0) + 1,
        }
        
        # Convert dates if provided
        if "dueAt" in update_data and update_data["dueAt"]:
            update_data["dueAt"] = datetime.fromisoformat(
                update_data["dueAt"].replace("Z", "+00:00")
            )
        elif "dueAt" in update_data:
            update_data["dueAt"] = None
        
        if "recurrenceEndDate" in update_data and update_data["recurrenceEndDate"]:
            update_data["recurrenceEndDate"] = datetime.fromisoformat(
                update_data["recurrenceEndDate"].replace("Z", "+00:00")
            )
        elif "recurrenceEndDate" in update_data:
            update_data["recurrenceEndDate"] = None
        
        # Update shared status based on sharedWith array
        if "sharedWith" in update_data:
            update_data["isShared"] = bool(
                update_data["sharedWith"] and len(update_data["sharedWith"]) > 0
            )
        
        reminder_ref.update(update_data)
        
        updated_reminder = {
            "id": reminder_id,
            **reminder,
            **update_data,
        }
        
        if format_for_llm:
            return format_reminder_for_llm(updated_reminder)
        
        return updated_reminder
    except Exception as error:
        raise Exception(f"Error updating reminder: {str(error)}")


async def delete_reminder(reminder_id: str, user_id: str) -> Dict[str, Any]:
    """
    Delete a reminder
    
    Args:
        reminder_id: Reminder ID
        user_id: User email or ID (must be owner)
        
    Returns:
        Success message
    """
    try:
        db = get_firestore()
        reminder_ref = db.collection("shared_tasks").document(reminder_id)
        doc = reminder_ref.get()
        
        if not doc.exists:
            raise Exception("Reminder not found")
        
        reminder = doc.to_dict()
        
        # Verify ownership
        if reminder.get("ownerId") != user_id:
            raise Exception("Access denied: You can only delete your own reminders")
        
        # Soft delete
        reminder_ref.update({
            "deleted": True,
            "deletedAt": datetime.now(),
            "lastModifiedBy": user_id,
            "version": (reminder.get("version") or 0) + 1,
        })
        
        return {
            "success": True,
            "message": f'Reminder "{reminder.get("title")}" deleted successfully',
            "reminderId": reminder_id,
        }
    except Exception as error:
        raise Exception(f"Error deleting reminder: {str(error)}")


async def complete_reminder(reminder_id: str, user_id: str) -> Dict[str, Any]:
    """
    Mark a reminder as completed
    
    Args:
        reminder_id: Reminder ID
        user_id: User email or ID (must be owner or has access)
        
    Returns:
        Completed reminder object
    """
    try:
        db = get_firestore()
        reminder_ref = db.collection("shared_tasks").document(reminder_id)
        doc = reminder_ref.get()
        
        if not doc.exists:
            raise Exception("Reminder not found")
        
        reminder = doc.to_dict()
        
        # Verify access
        if (
            reminder.get("ownerId") != user_id
            and user_id not in (reminder.get("sharedWith") or [])
        ):
            raise Exception(
                "Access denied: You do not have permission to complete this reminder"
            )
        
        now = datetime.now()
        
        reminder_ref.update({
            "isCompleted": True,
            "completedAt": now,
            "lastModifiedBy": user_id,
            "version": (reminder.get("version") or 0) + 1,
        })
        
        completed_reminder = {
            "id": reminder_id,
            **reminder,
            "isCompleted": True,
            "completedAt": now,
            "lastModifiedBy": user_id,
            "version": (reminder.get("version") or 0) + 1,
        }
        
        return format_reminder_for_llm(completed_reminder)
    except Exception as error:
        raise Exception(f"Error completing reminder: {str(error)}")
